src/NLog.Extensions.Logging/Extensions/ConfigureExtensions.cs

```diff
@@ -177,9 +177,27 @@ public static ILoggingBuilder AddNLog(this ILoggingBuilder builder, Func<IServic
             return builder;
         }
 
-        private static void AddNLogLoggerProvider(IServiceCollection services, IConfiguration configuration, NLogProviderOptions options, Func<IServiceProvider, IConfiguration, NLogProviderOptions, NLogLoggerProvider> factory)
+        private static void AddNLogLoggerProvider(ILoggingBuilder builder, IConfiguration configuration, NLogProviderOptions options, Func<IServiceProvider, IConfiguration, NLogProviderOptions, NLogLoggerProvider> factory)
         {
-            services.TryAddEnumerable(ServiceDescriptor.Singleton<ILoggerProvider, NLogLoggerProvider>(serviceProvider => factory(serviceProvider, configuration, options)));
+            var services = builder.Services;
+            var sharedFactory = factory;
+
+            if ((options ?? NLogProviderOptions.Default).ReplaceLoggerFactory)
+            {
+                NLogLoggerProvider singleInstance = null;   // Ensure that registration of ILoggerFactory and ILoggerProvider shares the same single instance
+                sharedFactory = (provider, cfg, opt) => singleInstance ?? (singleInstance = factory(provider, cfg, opt));
+
+                builder.ClearProviders();   // Cleanup the existing LoggerFactory, before replacing it with NLogLoggerFactory
+                services.Replace(ServiceDescriptor.Singleton<ILoggerFactory, NLogLoggerFactory>(serviceProvider => new NLogLoggerFactory(sharedFactory(serviceProvider, configuration, options))));
+            }
+
+            services.TryAddEnumerable(ServiceDescriptor.Singleton<ILoggerProvider, NLogLoggerProvider>(serviceProvider => sharedFactory(serviceProvider, configuration, options)));
+
+            if ((options ?? NLogProviderOptions.Default).ResetLoggerProviderFilter)
+            {
+                // Will forward all messages to NLog if not specifically overridden by user
+                builder.AddFilter<NLogLoggerProvider>(null, Microsoft.Extensions.Logging.LogLevel.Trace);
+            }
```

Comment:
is this duplicate code?

Reply:
I don't mind duplicate MEL-boilerplate-code, since they live in two different assemblies.

Reply:
Also there is a small difference, as UseNLog happens outside AddLogging, and so is more clean. Since changing the carpet (LoggerFactory) while standing in the room (AddLogging) is not perfect.

Reply:
> Also there is a small difference

Thanks for proving the issue with code duplication ;) 
There is a small difference and I can't see that in the code. That really hurt maintenance.